src/dcegm/solve.py

```diff
@@ -241,27 +241,41 @@ def backwards_induction(
         for child_state in possible_child_states:
 
             child_state_index = state_indexer[tuple(child_state)]
-            # We could parralelize here also over the savings grid!
-            # We aggregate here already over the income shocks!
-
-            (
-                marginal_utilities[child_state_index, :],
-                max_expected_values[child_state_index, :],
-            ) = get_child_state_marginal_util_and_exp_max_value(
-                exogenous_savings_grid,
-                income_shock_draws,
-                income_shock_weights,
-                child_state,
-                state_indexer,
-                state_space,
-                taste_shock_scale,
-                policy_array,
-                value_array,
-                compute_next_wealth_matrices,
-                compute_marginal_utility,
-                compute_value,
-                get_state_specific_choice_set,
-            )
+
+            for savings_index in range(len(exogenous_savings_grid)):
+                saving = exogenous_savings_grid[savings_index]
+
+                marginal_utility_weighted = 0
+                max_exp_value_weighted = 0
+                for shock_index in range(len(income_shock_draws)):
+                    income_shock = income_shock_draws[shock_index]
+                    income_shock_weight = income_shock_weights[shock_index]
+                    (
+                        marginal_util_weighted_shock,
+                        max_exp_value_weighted_shock,
+                    ) = get_child_state_marginal_util_and_exp_max_value(
+                        saving,
+                        income_shock,
+                        income_shock_weight,
+                        child_state,
+                        state_indexer,
+                        state_space,
+                        taste_shock_scale,
+                        policy_array,
+                        value_array,
+                        compute_next_period_wealth,
+                        compute_marginal_utility,
+                        compute_value,
+                        get_state_specific_choice_set,
+                    )
+                    marginal_utility_weighted += marginal_util_weighted_shock
+                    max_exp_value_weighted += max_exp_value_weighted_shock
+                marginal_utilities[
```

Comment:
Maybe just initialize the array with zeros and then add to each entry the calculated weighted marginal_utilities/max_exp_value